maps-compose/src/main/java/com/google/maps/android/compose/GoogleMap.kt

```diff
@@ -123,107 +137,224 @@ public fun GoogleMap(
-                ) {
-                    currentContent?.invoke()
+
+    val context = LocalContext.current
+
+    // Debug stuff
+    val debugCompositionId = remember { compositionCounter++ }
+    var debugMapId: Int? by remember { mutableStateOf(null) }
+    var debugIsMapReused: Boolean? by remember { mutableStateOf(null) }
+
+    /**
+     * Create and apply the [content] compositions to the map +
+     * dispose the [Composition] when the parent composable is disposed.
+     * */
+    fun CoroutineScope.launchComposition(mapView: MapView): Job {
+        mapView.log("Creating composition...")
+
+        val mapCompositionContent: @Composable () -> Unit = {
+            MapUpdater(
+                mergeDescendants = mergeDescendants,
+                contentDescription = currentContentDescription,
+                cameraPositionState = currentCameraPositionState,
+                contentPadding = currentContentPadding,
+                locationSource = currentLocationSource,
+                mapProperties = currentMapProperties,
+                mapUiSettings = currentUiSettings,
+            )
+
+            MapClickListenerUpdater()
+
+            CompositionLocalProvider(
+                LocalCameraPositionState provides currentCameraPositionState,
+            ) {
+                currentContent?.invoke()
+            }
+        }
+
+        return launch(start = CoroutineStart.UNDISPATCHED) {
+            val composition = mapView.createComposition(mapClickListeners, parentComposition).apply {
+                setContent(mapCompositionContent)
+            }
+
+            try {
+                awaitCancellation()
+            } finally {
+                mapView.log("Disposing composition...")
+                composition.dispose()
+            }
+        }
+    }
+
+    var isCompositionSet by remember { mutableStateOf(false) }
+    val mapUpdaterScope = rememberCoroutineScope()
+
+    AndroidView(
+        modifier = modifier,
+        factory = {
+            debugIsMapReused = false
+            MapView(context, googleMapOptionsFactory()).also { mapView ->
+                mapView.log("Creating MapView")
+                mapView.registerAndSaveNewComponentCallbacks(context)
+
+                fun log(msg: String) = mapView.log(msg)
+
+                val lifecycleObserver = object : LifecycleEventObserver {
+                    private var currentLifecycleState: Lifecycle.State = Lifecycle.State.INITIALIZED
+
+                    override fun onStateChanged(source: LifecycleOwner, event: Lifecycle.Event) {
+                        if(event == Lifecycle.Event.ON_DESTROY) {
+                            // Only destroy from onRelease. Move down to CREATED state.
+                            moveToLifecycleState(Lifecycle.State.CREATED)
+                            return
+                        }
+                        moveToLifecycleState(event.targetState)
+                    }
+
+                    @Synchronized
```

Comment:
Do you see a specific reason requiring this to be synchronized? I think all relevant ops here operate on main thread.

Reply:
Yes, it's probably not needed. Will remove that.